components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -48,16 +53,129 @@ AssetDiscoveryManager::GetAssetDiscoverySupportedChains() {
   return *asset_discovery_supported_chains;
 }
 
-void AssetDiscoveryManager::DiscoverAssets(
+void AssetDiscoveryManager::DiscoverSolAssets(
+    const std::vector<std::string>& account_addresses,
+    bool triggered_by_accounts_added) {
+  if (account_addresses.empty()) {
+    CompleteDiscoverAssets(
+        mojom::kSolanaMainnet, std::vector<mojom::BlockchainTokenPtr>(),
+        absl::nullopt, l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS),
+        triggered_by_accounts_added);
+    return;
+  }
+
+  // TODO(nvonpentz): When custom networks are supported, we need to check
+  // that the active network is our own that supports this RPC call.
+
+  const auto barrier_callback = base::BarrierCallback<std::vector<std::string>>(
+      account_addresses.size(),
+      base::BindOnce(&AssetDiscoveryManager::MergeDiscoveredSolanaAssets,
+                     weak_ptr_factory_.GetWeakPtr(),
+                     triggered_by_accounts_added));
+  for (const auto& account_address : account_addresses) {
+    json_rpc_service_->GetSolanaTokenAccountsByOwner(
+        account_address,
+        base::BindOnce(&AssetDiscoveryManager::OnGetSolanaTokenAccountsByOwner,
+                       weak_ptr_factory_.GetWeakPtr(),
+                       std::move(barrier_callback)));
+  }
+}
+
+void AssetDiscoveryManager::OnGetSolanaTokenAccountsByOwner(
+    base::OnceCallback<void(std::vector<std::string>)> barrier_callback,
+    const std::vector<absl::optional<SolanaAccountInfo>>& token_accounts,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::SolanaProviderError::kSuccess || token_accounts.empty()) {
+    std::move(barrier_callback).Run(std::vector<std::string>());
+    return;
+  }
+
+  // Add each token account to the all_discovered_contract_addresses list
+  std::vector<std::string> discovered_mint_addresses;
+  for (const auto& token_account : token_accounts) {
+    // Get the contract address
+    if (token_account.has_value()) {
+      // Decode Base64
+      const absl::optional<std::vector<uint8_t>> data =
+          base::Base64Decode(token_account->data);
+      if (data.has_value()) {
+        // Decode the address
+        const absl::optional<std::string> mint_address =
+            DecodeMintAddress(data.value());
+        if (mint_address.has_value()) {
+          // Add the contract address to the list
+          discovered_mint_addresses.push_back(mint_address.value());
+        }
+      }
+    }
+  }
+
+  std::move(barrier_callback).Run(discovered_mint_addresses);
+}
+
+void AssetDiscoveryManager::MergeDiscoveredSolanaAssets(
+    bool triggered_by_accounts_added,
+    const std::vector<std::vector<std::string>>&
```

Comment:
I'd suggest using `SolanaAddress` instead of `string` so code around becomes a bit more typesafe and clear